### PR TITLE
Add destroy time flag

### PR DESCRIPTION
**Pull request by syst0m**
# PR o'clock

## Description
Due to limitations with destroy-time provisioners, removing the module instance
from the configuration and running `terraform apply` will not run any destroy-time provisioners
associated with the module.

[Destroy-Time Provisioners](https://www.terraform.io/docs/provisioners/index.html#destroy-time-provisioners)
[19355](https://github.com/hashicorp/terraform/issues/19355)
[13549](https://github.com/hashicorp/terraform/issues/13549)

This PR adds an `enabled` flag, which can be toggled to set count on all resources to `0`,
thus allowing any destroy-time provisioners to run successfully.

### Checklist

- [X] Change added to CHANGELOG.md. All changes must be added and breaking changes and highlighted
- [X] CI tests are passing
- [X] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by max-rocket-internet** [comment]
A large PR 😅

**Comment by syst0m** [comment]
> A large PR

Yeah, if you know of a simpler way to do this, please let me know. :)

**Comment by barryib** [comment]
Hmmm, 2 remarques:

1. I would prefer `create_resources` instead of `enable`.
2. This is good option, but how can we handle migration ? It sounds like this will change all resource identifiers. With the `count` usage, resources are now lists.

**Comment by max-rocket-internet** [comment]
I still don't get the point here, there is no destroy-time provisioners in this module?

**Comment by barryib** [comment]
Aggreed with @max-rocket-internet.

@syst0m can you please explain your use case ? What are you trying to achieve/solve ?

**Comment by syst0m** [comment]
> I still don't get the point here, there is no destroy-time provisioners in this module?

I define the destroy-time provisioner, from my module, when calling the eks module.
It didn't make sense to include it in this module, since usecases might differ.
My destroy-time provisioner checks for public load balancers, and exits on finding any.
If there aren't any, it deletes all objects in the k8s cluster.
Also, due to a limitation in terraform, the destroy-time provisioners don't run if you 
remove the module block from the config, and run apply.
See: [Destroy-Time Provisioners](https://www.terraform.io/docs/provisioners/index.html#destroy-time-provisioners)

The described workaround requires using `count` on the resources,
hence this PR.

> Aggreed with @max-rocket-internet.
> 
> @syst0m can you please explain your use case ? What are you trying to achieve/solve ?

Please, see my comment above.
Also, there's a vpc related bug which manifests, when destroying the vpc resources created with this module.
If an associated EIP has tags, destroy fails with:
```AuthFailure: You do not have permission to access the specified resource.```
See: [#64](https://github.com/terraform-aws-modules/terraform-aws-vpc/issues/64)
&
[#2968](https://github.com/terraform-providers/terraform-provider-aws/issues/2968)

This could also be adressed with a destroy-time provisioner, by disassociating the EIPs
during destroy-time.

**Comment by syst0m** [comment]
There seems to be an issue with one of the outputs modification, I made in this PR.
I'm just looking into it.

```

[2019-11-06T10:57:42Z] module.habito-eks-skynet.aws_iam_role_policy_attachment.argocd_repo_server_can_decrypt_k8s_sops: Creating...
[2019-11-06T10:57:42Z] module.habito-eks-skynet.aws_iam_role_policy_attachment.argocd_repo_server_can_decrypt_k8s_sops: Creation complete after 1s [id=argocd-repo-server.argocd@skynet-20191106105742666500000011]
[2019-11-06T10:57:43Z]
[2019-11-06T10:57:43Z] Error: Error in function call
[2019-11-06T10:57:43Z]
[2019-11-06T10:57:43Z]   on ../modules/terraform-aws-eks/outputs.tf line 13, in output "cluster_certificate_authority_data":
[2019-11-06T10:57:43Z]   13:   value       = element(concat(flatten(aws_eks_cluster.this[0].certificate_authority[0].data), list("")), 0)
[2019-11-06T10:57:43Z]     |----------------
[2019-11-06T10:57:43Z]     | aws_eks_cluster.this[0].certificate_authority[0].data is "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUN5RENDQWJDZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRFNU1URXdOakV3TlRNeU5sb1hEVEk1TVRFd016RXdOVE15Tmxvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTHhOClNjS3htWnhySXo3RnI0YjFtN2dLWXpLKys0YXZUZFlPa0trKzUvS21rcTFESEdPYzFBK1l2azkydHVaTUxHRWEKZHdqa0xPR2RtRjhLOEVRTzRjV0IxZlh0VDBkL1EwTlNnWXdobXdhcTNFMnBHU1V5ei9mVTloL2JyTU9WcVhDcwpEdkFKcGJRK0xHaDlYWEJyR3I2NG0zZFlTQjI0N291YXAwZzZzc2pYdlJBVk9MUlFIdHdNVVRUYXRxcitJZ0l3Ck5XUWtaT0ZORzE2K3E1VW5jSkxrRnFkaXFVQlc5VW9pQk4wQmJKWW1iRFZnMVczb2wxdDJtc2NpZElTMGZwMEYKU2pNcG1rSWtkZzBBbUpYRXBNVkpiMXBjQ1NiK1Q3UUJEc2pmL2RtSzluVXFqbGxmckNKYXZOa0dXM0FSZTArRApIakQxaDdLNUpQcTVKVDZqVmtNQ0F3RUFBYU1qTUNFd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0RRWUpLb1pJaHZjTkFRRUxCUUFEZ2dFQkFIa21yV1VGRko1WXhqWXk2V0FBQzNOWmhwWDgKQzRXd2Nhb3BvMGphNm8wRTRZVnNTWWJTSWxiYytKT3FSSVBkVUtnQ0FDTGRXb04zZzVVOUZpU09CbkNQRVNFcgpyOXpYN2tMM3JJZHhjamE3eFN4M1MwMjZRNnlkWTJuK1dOVnY5Y1ZJQzlGRXc1aU05RERuT3R5U0hoVHpTQnh0CnVxaGUvRWFGUi9rcldmKy9SMGFLS0U1TFdSc0wzT3dCUENqSlRuSkQ2K04rcGg0Y3QyR1NHNUNvWWw4elJhL1gKOXBvV2s1blFJdXIrRGZvalAwb2lFZ3JlZzY5c29JaWhoazk3TlhEa3JVeEhWeDJCclVackxNbXZMalpaUUhmaQpIMWFPZXlFcm5zNEczN1cvdGxJNENuRmwwaUs1cW9SRFk2VUM3Zk9OcitoaGhEbE51UkNJNno3ZE13ND0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo="
[2019-11-06T10:57:43Z]
[2019-11-06T10:57:43Z] Call to function "flatten" failed: can only flatten lists, sets and tuples.
[2019-11-06T10:57:43Z]
[2019-11-06T10:57:43Z] [terragrunt] 2019/11/06 10:57:43 Hit multiple errors:
[2019-11-06T10:57:43Z] exit status 1
```

**Comment by max-rocket-internet** [comment]
> I define the destroy-time provisioner, from my module, when calling the eks module

OK.

It's a pretty large change that seems quite niche. It would also require everyone to make manual changes to their state since resources will be renamed like @barryib mentioned. I'm not super keen on merging this unless there's huge support from users.

> My destroy-time provisioner checks for public load balancers, and exits on finding any.

Surely you can achieve this another way?

**Comment by syst0m** [comment]
> It's a pretty large change that seems quite niche. It would also require everyone to make manual changes to their state since resources will be renamed like @barryib mentioned. I'm not super keen on merging this unless there's huge support from users.

Regarding the migration, that part is tricky I agree.
On the point of this being a niche usecase, 
I would expect people to try to recreate their clusters on a regular basis.

> > My destroy-time provisioner checks for public load balancers, and exits on finding any.
> 
> Surely you can achieve this another way?

There is an issue with running the destroy-time provisioner when using `terraform destroy`.
I'll open a separate issue for it.

**Comment by dpiddockcmp** [comment]
>     2. How can we handle migration ? It sounds like this will change all resource identifiers. With the `count` usage, resources are now lists.

Hi @barryib , you're correct in saying that all resources are now lists but there is no migration needed for existing users. Terraform does some internal magic when moving between a single resource and a count of 1 😄

File: outputs.tf [hunk 1]
```diff
@@ -1,26 +1,26 @@
...
 output "cluster_arn" {
   description = "The Amazon Resource Name (ARN) of the cluster."
-  value       = aws_eks_cluster.this.arn
+  value       = element(concat(aws_eks_cluster.this.*.arn, list("")), 0)
 }
 
 output "cluster_certificate_authority_data" {
   description = "Nested attribute containing certificate-authority-data for your cluster. This is the base64 encoded certificate data required to communicate with your cluster."
-  value       = aws_eks_cluster.this.certificate_authority[0].data
+  value       = element(concat(aws_eks_cluster.this.*.certificate_authority[0].data, list("")), 0)
```
[the remark this answers is not in the document]
Comment (syst0m) [reply outputs.tf]: Thanks for your suggestion.
I managed to solve this in another way, but this solution seems more elegant.

File: variables.tf [hunk 2]
```diff
@@ -293,3 +293,9 @@ variable "attach_worker_cni_policy" {
   type        = bool
   default     = true
 }
+
+variable "enabled" {
+  description = "Destroy-time flag. If set to `false`, the count will be set to `0` on all resources."
```
[the remark this answers is not in the document]
Comment (syst0m) [reply variables.tf]: Many thanks for your suggestions, and spotting and resolving some issues with the PR, @dpiddockcmp

**Comment by max-rocket-internet** [comment]
> On the point of this being a niche usecase, I would expect people to try to recreate their clusters on a regular basis.

Why? If you have a large complex environment, recreating a whole cluster is simply not possible unless it's some sort of DR situation.

I still don't get it though. All this just to check for public load balancers during deploy? If the ELB belongs to a service then delete the service and k8s will delete the ELB for you. Just clean up services before deleting the cluster. This will also clean up security groups 🙂

**Comment by syst0m** [comment]
> > On the point of this being a niche usecase, I would expect people to try to recreate their clusters on a regular basis.
> 
> Why? If you have a large complex environment, recreating a whole cluster is simply not possible unless it's some sort of DR situation.

When you build on top of the module which provisions the cluster resources,
I think the best way to detect cyclic dependencies, inconsistencies, bootstrapping issues
and things that break automated provisioning, 
is to have a pipeline which continuously recreates the cluster,
and all it's addon resources on a staging environment.

> I still don't get it though. All this just to check for public load balancers during destroy? If the ELB >belongs to a service then delete the service and k8s will delete the ELB for you. Just clean up >services before deleting the cluster. This will also clean up security groups

It's also about being able to destroy the cluster from a pipeline, by using `terraform apply`.
Please see linked issues in original post.
So, the goal here is to abort the destroy, if there are public load balancers.
The reasoning is that if there are public lbs, some of the services are still likely active and receiving 
traffic. Having this protects live services from accidental deletions.
As for cleaning up the services, the logical place to do this, is with a destroy-time provisioner,
depending on the cluster.
I hope this makes sense.

**Comment by jcpetruzza** [comment]
I'd suggest using `create_cluster` or `create_eks` instead of `enabled`, for consistency with other modules like [terraform-aws-alb](https://github.com/terraform-aws-modules/terraform-aws-alb#conditional-creation) (`create_alb`) or [terraform-aws-vpc](https://github.com/terraform-aws-modules/terraform-aws-vpc#conditional-creation) (`create_vpc`)

**Comment by syst0m** [comment]
Renamed the flag to `create_cluster`.
Also, included a documentation section on conditional creation, similar to the ones in the alb and vpc modules.

**Comment by barryib** [comment]
@syst0m @jcpetruzza @max-rocket-internet Is `create_cluster` the greatest name we can have ? To me this is ambiguous, because **cluster** means master nodes when you're dealing with AWS EKS. Even the terraform resource name is `aws_eks_cluster`. So when I see `create_cluster` I think that this only for the cluster resources.

I would suggest `create_resources` which is more generic or something else.

Just thinking.

**Comment by jcpetruzza** [comment]
@barryib Naming is hard! :smiley:

I find `create_resources` a bit vague, and I'd definitely have to refer to the docs to see what it does. That said, I think I'd have to do the same regardless of the name we give it, so it's fine by me as well!

My personal choice would be `create_eks`, only because there would be some consistency between modules, so one could end up with the rule: "`create_foo=false` in `terraform-aws-foo` is the way to disable the module"

**Comment by syst0m** [comment]
Probably makes sense then to use the same convention as other modules, as in `create_eks`.

**Comment by barryib** [comment]
Let's go with `create_eks` then. @max-rocket-internet any thoughts ?

**Comment by max-rocket-internet** [comment]
> Let's go with create_eks then

OK cool 👍

I'll hopefully have time to review this PR this week.

**Comment by TBeijen** [comment]
> Hi @barryib , you're correct in saying that all resources are now lists but there is no migration needed for existing users. Terraform does some internal magic when moving between a single resource and a count of 1 

Wasn't aware of that either. Can confirm, created EKS cluster with latest module version, then switched to PR source. No proposed changes.

Another use case for this feature is having a `.tfvars` based project layout spanning multiple AWS accounts (e.g. production and test) where not all clusters need to exist in all accounts.

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,7 +11,7 @@ project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Added
 
-- Test against minimum versions specified in `versions.tf` (by @dpiddockcmp)
```
Comment (max-rocket-internet) [inline CHANGELOG.md]: Why are you removing change log entry?

File: README.md [hunk 4]
```diff
@@ -44,6 +44,19 @@ module "my-cluster" {
...
+## Conditional creation
+
+Sometimes you need to have a way to create EKS resources conditionally but Terraform does not allow to use `count` inside `module` block, so the solution is to specify argument `create_cluster`.
+
+```hcl
+# This cluster will not be created
+module "eks" {
+  source = "terraform-aws-modules/eks/aws"
+
+  create_cluster = false
```
Comment (max-rocket-internet) [inline README.md]: I think we switch this to `create_eks` as discussed

File: README.md [hunk 5]
```diff
@@ -44,6 +44,19 @@ module "my-cluster" {
   }
 }
 ```
+## Conditional creation
+
+Sometimes you need to have a way to create EKS resources conditionally but Terraform does not allow to use `count` inside `module` block, so the solution is to specify argument `create_cluster`.
```
Comment (max-rocket-internet) [inline README.md]: Switch to `create_eks` here also.

File: data.tf [hunk 6]
```diff
@@ -117,9 +118,9 @@ data "template_file" "userdata" {
 
   vars = merge({
     platform            = lookup(var.worker_groups[count.index], "platform", local.workers_group_defaults["platform"])
-    cluster_name        = aws_eks_cluster.this.name
-    endpoint            = aws_eks_cluster.this.endpoint
-    cluster_auth_base64 = aws_eks_cluster.this.certificate_authority[0].data
+    cluster_name        = aws_eks_cluster.this[count.index].name
```
Comment (max-rocket-internet) [inline data.tf]: Do we need `[count.index]` here? can we just have `[0]` for simplicity and consistency?

File: variables.tf [hunk 7]
```diff
@@ -293,3 +293,9 @@ variable "attach_worker_cni_policy" {
   type        = bool
   default     = true
 }
+
+variable "create_cluster" {
```
Comment (max-rocket-internet) [inline variables.tf]: Let's switch to `create_eks` as discussed

File: CHANGELOG.md (hunk 3 above)
Comment (syst0m) [reply CHANGELOG.md]: Sorry, this was a mistake.
I added back the entry.

File: data.tf (hunk 6 above)
Comment (syst0m) [reply data.tf]: I updated this, to be consistent with other vars.

**Comment by max-rocket-internet** [comment]
@syst0m I get these errors when running this against an existing cluster:

```
Error: Invalid index

  on .terraform/modules/eu01_stg/data.tf line 170, in data "template_file" "launch_template_userdata":
 170:     cluster_name        = aws_eks_cluster.this[count.index].name
    |----------------
    | aws_eks_cluster.this is tuple with 1 element
    | count.index is 1

The given key does not identify an element in this collection value.


Error: Invalid index

  on .terraform/modules/eu01_stg/data.tf line 171, in data "template_file" "launch_template_userdata":
 171:     endpoint            = aws_eks_cluster.this[count.index].endpoint
    |----------------
    | aws_eks_cluster.this is tuple with 1 element
    | count.index is 1

The given key does not identify an element in this collection value.


Error: Invalid index

  on .terraform/modules/eu01_stg/data.tf line 172, in data "template_file" "launch_template_userdata":
 172:     cluster_auth_base64 = aws_eks_cluster.this[count.index].certificate_authority[0].data
    |----------------
    | aws_eks_cluster.this is tuple with 1 element
    | count.index is 1

The given key does not identify an element in this collection value.
```

**Comment by max-rocket-internet** [comment]
If I set those to `aws_eks_cluster.this[0].xx` then I get another error:

```
Error: Invalid index

  on .terraform/modules/eu01_stg/workers_launch_template.tf line 405, in resource "aws_iam_instance_profile" "workers_launch_template":
 405:   name_prefix = aws_eks_cluster.this[count.index].name
    |----------------
    | aws_eks_cluster.this is tuple with 1 element
    | count.index is 1
```

**Comment by TBeijen** [comment]
@syst0m If you don't mind adding me as collaborator to your forked repo, I might be able to take a look at this the coming days.

**Comment by syst0m** [comment]
> @syst0m If you don't mind adding me as collaborator to your forked repo, I might be able to take a look at this the coming days.

Sure, thanks for the help, @TBeijen  :)



> If I set those to `aws_eks_cluster.this[0].xx` then I get another error:
> 
> ```
> Error: Invalid index
> 
>   on .terraform/modules/eu01_stg/workers_launch_template.tf line 405, in resource "aws_iam_instance_profile" "workers_launch_template":
>  405:   name_prefix = aws_eks_cluster.this[count.index].name
>     |----------------
>     | aws_eks_cluster.this is tuple with 1 element
>     | count.index is 1
> ```

I will try to reproduce it, @max-rocket-internet

**Comment by TBeijen** [comment]
I managed to reproduce the same errors, first `data.tf:170-172`, then `workers_launch_template.tf:405`. Key to reproduce was having more than 1 worker group.

At several points the created cluster was referenced via `aws_eks_cluster.this[count.index]` which fails if iterating over more than 1 resource (which happens if having >1 worker groups).

I replaced all occurences of `aws_eks_cluster.this[count.index]` with `aws_eks_cluster.this[0]`. This should be safe as all the affected resources are based on `var.create_eks` and there can be max. 1 eks cluster:

```
# /cluster.tf
resource "aws_eks_cluster" "this" {
  count                     = var.create_eks ? 1 : 0
```

What I tested:
* Create cluster using v7.0.0, having 2 worker groups based on launch templates.
* Switching module to https://github.com/TBeijen/terraform-aws-eks/tree/a4317715531c7cecc226324160c33b0105c90dd4 -> no proposed changes.

@syst0m Can you add me as collaborator, then I'll push to this PR? (Alternatively, if fine by @max-rocket-internet, I can create a new PR)

**Comment by syst0m** [comment]
Great, thansk for the help, @TBeijen ! :)

**Comment by TBeijen** [comment]
@max-rocket-internet I merged in latest master, including the aws_auth changes (which works very nice btw.).

What would be needed to move this PR forward? (besides spare time of reviewers obviously)

Tagged it `v8.*.*` in Changelog btw, as I assume the breaking aws_auth change will result in a major version bump.

**Comment by max-rocket-internet** [comment]
OK I did another test from `examples/basic` with `create_eks = false` and `terraform apply` gives these errors:

```
Error: name must not be empty

  on main.tf line 26, in data "aws_eks_cluster" "cluster":
  26: data "aws_eks_cluster" "cluster" {

Error: name must not be empty

  on main.tf line 30, in data "aws_eks_cluster_auth" "cluster":
  30: data "aws_eks_cluster_auth" "cluster" {
```

**Comment by TBeijen** [comment]
Ah, ok. Looks like the example needs some updating. I'll look into it and tweak some other docs/examples as well (iirc there's also some references to `kubectl` and `iam_authenticator` lingering which is no longer relevant with the introduction of kubernetes provider)

I use a wrapper module (which also sets up ingress ALBs, some route53 things, etc.) and there I've implemented it like this:

```
data "aws_eks_cluster" "eks_cluster" {
  count = var.enabled ? 1 : 0
  name  = module.eks_cluster.cluster_id
}

data "aws_eks_cluster_auth" "eks_cluster" {
  count = var.enabled ? 1 : 0
  name  = module.eks_cluster.cluster_id
}

provider "kubernetes" {
  host                   = element(concat(data.aws_eks_cluster.eks_cluster[*].endpoint, list("")), 0)
  cluster_ca_certificate = base64decode(element(concat(data.aws_eks_cluster.eks_cluster[*].certificate_authority.0.data, list("")), 0))
  token                  = element(concat(data.aws_eks_cluster_auth.eks_cluster[*].token, list("")), 0)
  load_config_file       = false
  version                = "~> 1.10"
}
```

So, this could leave the provider configured with empty host, cert. and token, which is no problem as it won't be invoked.

**Comment by TBeijen** [comment]
Done. 
The `element(concat(` being neccessary outside of the module is not overly elegant. Which made wonder: Can't the kubernetes provider be set up _inside_ the eks module? Would be happy to open an issue/PR for that.

**Comment by dpiddockcmp** [comment]
Configuring the k8s provider in the module is what the [GKE module ](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/auth.tf#L29..L34) does. But then you can't reuse that provider outside the module to configure other resources. Only an issue for more advanced users to resolve?

Plus Hashicorp recommends against configuring providers within modules. [docs](https://www.terraform.io/docs/configuration/modules.html#providers-within-modules)

File: README.md [hunk 8]
```diff
@@ -14,31 +14,38 @@ Read the [AWS docs on EKS to get connected to the k8s dashboard](https://docs.aw
 * You want to create an EKS cluster and an autoscaling group of workers for the cluster.
 * You want these resources to exist within security groups that allow communication and coordination. These can be user provided or created within the module.
 * You've created a Virtual Private Cloud (VPC) and subnets where you intend to put the EKS resources. The VPC satisfies [EKS requirements](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html).
-* If `manage_aws_auth = true`, it's required that both [`kubectl`](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl) (>=1.10) and [`aws-iam-authenticator`](https://github.com/kubernetes-sigs/aws-iam-authenticator#4-set-up-kubectl-to-use-authentication-tokens-provided-by-aws-iam-authenticator-for-kubernetes) are installed and on your shell's PATH.
```
Comment (max-rocket-internet) [inline README.md]: Ahh thanks for removing this

**Comment by max-rocket-internet** [comment]
> Plus Hashicorp recommends against configuring providers within modules

Yep ✅

@TBeijen 

> Done. The need for element(concat( being neccessary outside of the module is not overly elegant.

Thanks! But also sorry, this is quite ugly 😅 Can we undo all those last changes for the README and all the examples? Clean and clear is paramount for these parts.

As long as it's possible to `apply` with `create_eks = false` false *somehow,* we are good. I think this feature will be mainly for advanced users and they can work it out.

**Comment by TBeijen** [comment]
> Thanks! But also sorry, this is quite ugly 😅 Can we undo all those last changes for the README and all the examples? Clean and clear is paramount for these parts.

@max-rocket-internet Ugly it sure is! :) I could restore the current Readme and examples to their original state, also leaving out the `create_eks` variable. Then instead add  a 'conditional_create` folder under `/examples`? Seems a better plan indeed. 

Probably got some time to do it now. One sec.

**Comment by max-rocket-internet** [comment]
I that that would be great 🙂

**Comment by TBeijen** [comment]
Done, will also try to merge in the managed node group change from master. That will require some updates for sure.

**Comment by TBeijen** [comment]
@max-rocket-internet Merged in master. Didn't test _very_ extensively but runs fine on my current test project that has 2 clusters (one with `create_eks` is false).

Take not of that last commit. Change in `aws_auth.tf` already looked suspicious and planned to wipe all mapRoles from aws-auth.

File: aws_auth.tf [hunk 9]
```diff
@@ -53,7 +53,7 @@ resource "kubernetes_config_map" "aws_auth" {
   data = {
     mapRoles    = <<EOF
 ${join("", distinct(concat(data.template_file.launch_template_worker_role_arns.*.rendered, data.template_file.worker_role_arns.*.rendered)))}
-%{if var.map_roles == []}${yamlencode(var.map_roles)}%{endif}
+%{if var.map_roles != []}${yamlencode(var.map_roles)}%{endif}
```
Comment (max-rocket-internet) [inline aws_auth.tf]: FYI https://github.com/terraform-aws-modules/terraform-aws-eks/pull/611

**Comment by max-rocket-internet** [comment]
Wait:

> and planned to wipe all mapRoles from aws-auth.

Are you sure? @shanmugakarna was sure this was required...

**Comment by dpiddockcmp** [comment]
@shanmugakarna was only testing his use case of empty lists. Current master wipes out the roles map because the logic is wrong and Terraform is broken on complex lists. 🙈

**Comment by TBeijen** [comment]
> Wait:
> 
> > and planned to wipe all mapRoles from aws-auth.
> 
> Are you sure? @shanmugakarna was sure this was required...

Yes, definitely.

I already found it looking suspiciously, as it says "Only render MapRoles content... if you have defined _no_ map_roles".

`terraform plan` confirmed it, I have a project with several `map_roles` items. They would all be removed from the current cluster (which would be quite catastrophic).
 
I'm not exactly sure what problem case the if-condition aims to solve, but to me it looks like a true/false mix-up that slipped through (but breaks a lot). I'd be very careful to let current master version get into a release

**Comment by max-rocket-internet** [comment]
OK merged!

I think the auth issue is resolved now. Please doublecheck everything in master is working for you now correctly and open an issue or PR if now.

I'll also test it this week and make a new release as long as everything is OK.

**Comment by jurgenweber** [comment]
Hi @max-rocket-internet Long time.

Are we at the point where we could get a rls with this?

**Comment by barryib** [comment]
@jurgenweber There is already the PR https://github.com/terraform-aws-modules/terraform-aws-eks/pull/662 for a new release. It's actually on hold for https://github.com/terraform-aws-modules/terraform-aws-eks/pull/650 and it's pretty much finish. We'll release soon.

**Comment by max-rocket-internet** [comment]
Hey Jurgen, yes long time 😃 Hope you are safe 🔥🙏

Like @barryib says, in the next release which will happen in the next couple of days. We've had quite a few very large changes recently and a lot of discussion about how to handle the complexity going forwards. This has slowed things down. Also Christmas and NY etc.

**Comment by jurgenweber** [comment]
Yeah, we made it out of the fires fine..... Now the next bit. Sorry, only came across this again today.

Thanks team

**Pull request merged**